### PR TITLE
implementation methods in StreamPractice and CandidateValidator

### PR DESCRIPTION
**Pull request by BusyhinD**

### CODE REVIEW

File: src/main/java/practice/CandidateValidator.java [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+    private static final int INDEX_TO = 1;
+    private static final String REQUIRED_PRESIDENT_NATIONALITY = "Ukrainian";
+
+    @Override
+    public boolean test(Candidate candidate) {
+        return candidate.isAllowedToVote()
+                && candidate.getAge() >= REQUIRED_PRESIDENT_AGE
+                && candidate.getNationality().equals(REQUIRED_PRESIDENT_NATIONALITY)
+                && Integer.parseInt(candidate.getPeriodsInUkr().split("-")[INDEX_TO])
+                - Integer.parseInt(candidate.getPeriodsInUkr().split("-")[INDEX_SINCE])
```
Comment (Rommelua) [inline src/main/java/practice/CandidateValidator.java:19]: better create local var for the period to make this easier to read

File: src/main/java/practice/StreamPractice.java [hunk 2]
```diff
@@ -75,6 +115,11 @@ public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
      * parametrized with Candidate in CandidateValidator.
      */
     public List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
+        Predicate<Candidate> candidatePredicate = new CandidateValidator();
+        return candidates.stream()
+                .filter(candidatePredicate)
```
Comment (Rommelua) [inline src/main/java/practice/StreamPractice.java:120]:
```suggestion
                .filter(new CandidateValidator())
```

File: src/main/java/practice/StreamPractice.java [hunk 3]
```diff
@@ -75,6 +115,11 @@ public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
      * parametrized with Candidate in CandidateValidator.
      */
     public List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
+        Predicate<Candidate> candidatePredicate = new CandidateValidator();
```
Comment (Rommelua) [inline src/main/java/practice/StreamPractice.java:118]:
```suggestion
```

File: src/main/java/practice/CandidateValidator.java [hunk 4]
```diff
@@ -1,5 +1,24 @@
...
+    private static final int INDEX_SINCE = 0;
+    private static final int INDEX_TO = 1;
+    private static final String REQUIRED_PRESIDENT_NATIONALITY = "Ukrainian";
+
+    @Override
+    public boolean test(Candidate candidate) {
+        int sincePeriodInUkr = Integer.parseInt(
+                candidate.getPeriodsInUkr().split("-")[INDEX_SINCE]);
+        int toPeriodInUkr = Integer.parseInt(
+                candidate.getPeriodsInUkr().split("-")[INDEX_TO]);
```
Comment (aaananas) [inline src/main/java/practice/CandidateValidator.java:18]:
```suggestion
        String[] slitPeriod = candidate.getPeriodsInUkr().split("-");
        int sincePeriodInUkr = Integer.parseInt(
                splitPeriod[INDEX_SINCE]);
        int toPeriodInUkr = Integer.parseInt(
                splitPeriod)[INDEX_TO]);
```

File: src/main/java/practice/StreamPractice.java [hunk 5]
```diff
@@ -14,7 +21,14 @@ public class StreamPractice {
      * "Can't get min value from list: < Here is our input 'numbers' >"
      */
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers.stream()
+                .map(s -> s.split(","))
+                .flatMap(Arrays::stream)
```
Comment (aaananas) [inline src/main/java/practice/StreamPractice.java:26]: use just 1 flatMap here 
```suggestion
                .flatMap(s -> Arrays.stream(s.split(",")))
```

File: src/main/java/practice/StreamPractice.java [hunk 6]
```diff
@@ -59,7 +86,12 @@ public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
      * return the names of all cats whose owners are women from `femaleAge` years old inclusively.
      */
     public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> p.getSex() == Person.Sex.WOMAN && p.getAge() >= femaleAge)
+                .map(Person::getCats)
+                .flatMap(Collection::stream)
```
Comment (aaananas) [inline src/main/java/practice/StreamPractice.java:92]: use just 1 flatMap here

**Review by aaananas — CHANGES_REQUESTED** [review]

**Review by aaananas — APPROVED** [review]

**Pull request closed without merging**